docs/language-server/protocol-language-server.md

```diff
@@ -1124,8 +1126,43 @@ A representation of a change to a text file at a given position.
 
 ```typescript
 interface TextEdit {
+  /** The range of text in a text file. */
   range: Range;
+
+  /** The change to a text file. */
   text: String;
+
+  /** The update type. When not set defaults to `FileEditKind`. */
+  kind?: EditKind;
+}
+```
+
+### EditKind
```

Comment:
I would suggest that instead of adding a `EditKind` we should have two flags, that specify the follow-up behaviour of the edit.
`recompile: bool` and `re-execute: bool`. 
This would mean that it is the IDEs responsibility to decide which of these is appropriate for a given edit or sequence of edits. The docs should mention the intended use case (.e.g., edits to metadata). 

This avoids inconsistent edits, where the `EditKind` is incompatible with the actual edit. Presently, the API happily accepts an update to any part of thee file labelled as `MetadataEditKind`, which doesn't seem to make sense. 
While of course someone can still set the flags wrong, the API would be more consistent in itself.

For the same reason, I would suggest creating a new API endpoint for `ValueUpdates`, that also does not allow inconstant data to be provided.

Something like 
```
type ExpressionUpdate {
    expressionId: ExpressionId;
    new_expression: String;
}
```
This update would always be consistent.

Reply:
If IDE can take control of the execution, then the new endpoint makes sense. The one downside is that with this design IDE needs to make two requests for every slider update instead of one - one editing the expression, and second setting the value. Do you think this tradeoff is ok?

Reply:
> The one downside is that with this design IDE needs to make two requests for every slider update instead of one - one editing the expression, and second setting the value.
Can you give an example? I might be missing something here.

Reply:
> If IDE can take control of the execution, then the new endpoint makes sense. 

My main concern is, that the current proposal allows for nonsense to be submitted. In a perfect world, I guess the engine should be able to detect whether a change requires what kinds of follow-up action, but without that, I think we are better off to have a more explicit API.

Reply:
> Can you give an example? I might be missing something here.

For example, we have a slider on node allowing to change the value in the range [0, 100]
- User moves the slider from 18 to 19
- IDE sends `applyEdit` request to change the program text
- IDE sends `setExpression` request to set the node value to 99

Reply:
Right. `setExpression` changes not only the value, but also the underlying text. It is desirable to keep text and AST consistent. There are three options:
- combined `applyEdit` message where IDE requests both - the text change as well as _hints_ the _literal change_
- two messages from the IDE `applyEdit execute=false` and `setExpression`
-  one message from IDE to `setExpression` and _one message from the engine_ to the IDE that the code has been modified

There are various pros/cons to each option. Some favor consistency, some favor correctness.

Reply:
> one message from IDE to setExpression and one message from the engine to the IDE that the code has been modified

I like this one. We can even skip the Engine->IDE notification assuming that `setExpression` always changes the source.
The downside is that we can lose consistency because we don't send `SHA` before the change and `SHA` after, like we do in `applyEdit`.
Although we can add `oldVersion` and `newVersion` with `SHA` hashes to the `setExpression`.
But assuming that we want to enable the optimization with sending metadata, I'd propose the following:

Add `execute` flag to the `text/applyEdit` message indicating whether or not the program should be re-executed after applying the change.
```typescript
interface ApplyEdit {
  edit: FileEdit;
  /** New flag controlling the program execution. `true` if not set (current behavior). */
  execute?: boolean;
}
```
Create the new command `text/setExpression` that **sets** the new expression value *and* **modifies** the source code.
```typescript
interface SetExpression {
  /** Edited expression id. */
  expressionId: ExpressionId;
  /** The new value of the node. */
  expressionValue: string;
  /** Edited file. */
  path: Path;
  /** The sha hash of the file before setting the the new value to the node. */
  oldVersion: SHA3-224;
  /** The sha hash of the file after setting the the new value to the node. */
  newVersion: SHA3-224;
}